papers/joss/2025/paper.md

```diff
@@ -0,0 +1,214 @@
+Also, `ReProspect` goes significantly beyond runtime analysis and introduces new binary analysis functionalities to inspect machine code for expected instruction sequence patterns.
+
+# Key features
+
+`ReProspect` is organized into three main components:
+API tracing and kernel profiling for runtime analysis,
+and binary tools for static analysis (\ref{fig:overview}).
+
+Each component is designed to allow the entire analysis workflow to be encapsulated into a concise Python script.
+This includes launching the underlying analysis tool,
+collecting the output into Python data structures,
+and performing the subsequent analysis.
+
+![Overview of `ReProspect`.\label{fig:overview}](overview.svg){width="12cm"}
+
+## Runtime analysis
+
+The `ReProspect` `Command` and `Session` classes streamline launching
+NSight Systems and Nsight Compute to collect
+a focused set of metrics most relevant for the analysis.
+The collected data are gathered in a `Report`, queryable by NVTX range annotations [@nvtx],
+readily amenable to test assertions.
+
+To avoid unnecessary re-runs,
+`ReProspect` provides a `Cacher` that can serve the `Report` from a database.
+
+## Static analysis
+
+`ReProspect` provides a set of tools for extracting and analysing the content of CUDA binaries (CUBIN).
+
+The `CuObjDump` and `NVDisasm` classes drive and parse the output of the underlying CUDA utilities [@binary2025]
+to retrieve the SASS code and resource usage of kernels (*e.g.* register count, constant memory).
+
+The `ReProspect` `ELF` class can extract complementary information by decoding ELF-formatted sections,
+such as the symbol table, the toolchain (from the *.note.nv.tkinfo* section) or
+kernel attributes (*e.g.* launch bounds, from the *.nv.info.\<kernel\>* section) [@hayes-2019-decoding-cubin].
+
+Beyond data extraction, `ReProspect` provides an extensible framework for
+inspecting machine code for expected instruction sequence patterns.
+This framework is structured as a hierarchy of matchers.
+At the lowest level, matchers analyse instruction opcodes, modifiers, and operands.
+These instruction matchers can be composed into instruction sequence matchers,
+enabling the identification of more intricate patterns.
+
+One of the key challenges with SASS matching is the evolution of the CUDA instruction set and compiler toolchains.
+`ReProspect` addresses this challenge by developing matchers that express higher-level intents
+(*e.g.*, `LoadGlobalMatcher` for a load from global memory),
+with their implementation then handling the architecture- and version-specific details.
+This hierarchical design—spanning operand, instruction, and sequence matchers—
+enables robust, composable matching across multiple CUDA architectures and compiler versions
+while abstracting away low-level details.
+
+`ReProspect` already implements matchers for common operations
+such as load, store, atomic, or half-precision arithmetic.
```

Comment:
Could be deleted. Because it's covered by the previous sentences, by the code snippet, and by the desul use case.